From 0a42ea0339909210f4dba2ab652bbd4cd8724b8f Mon Sep 17 00:00:00 2001
From: Venkataramana Kotakonda <venkataramana.kotakonda@intel.com>
Date: Mon, 24 Aug 2015 21:25:05 +0530
Subject: [PATCH] USBC-PD: Data role swap changes to dpm

As per pd spec, data role swap state is not required as data role
can change directly from UFP to DFP or from DFP to UFP.
Hence removed DATA_ROLE_SWAP changes in DPM and handled role swap
accordingly.

Change-Id: Ie1528cfbfa6dc8a2a0dd3cf35f72de29b77c3335
Tracked-On: https://jira01.devtools.intel.com/browse/GMINL-15690
Signed-off-by: Venkataramana Kotakonda <venkataramana.kotakonda@intel.com>
---
 drivers/usb/typec/pd/devpolicy_mgr.c | 27 +++++++++++++--------------
 1 file changed, 13 insertions(+), 14 deletions(-)

diff --git a/drivers/usb/typec/pd/devpolicy_mgr.c b/drivers/usb/typec/pd/devpolicy_mgr.c
index 95fe0bfdbd50..1f11a2f5f070 100644
--- a/drivers/usb/typec/pd/devpolicy_mgr.c
+++ b/drivers/usb/typec/pd/devpolicy_mgr.c
@@ -439,29 +439,25 @@ static void dpm_update_data_role(struct devpolicy_mgr *dpm,
 	cur_drole = dpm->cur_drole;
 	if (cur_drole == drole)
 		goto drole_err;
+
 	switch (drole) {
-	case DATA_ROLE_SWAP:
+	case DATA_ROLE_UFP:
 		if (cur_drole == DATA_ROLE_DFP) {
 			/* Role swap from DFP to UFP, Send DFP disconnect */
-			cbl_type = "USB-Host";
-			cbl_state = CABLE_DETACHED;
-		} else if (cur_drole == DATA_ROLE_UFP) {
-			/* Role swap from UFP to DFP, Send UFP disconnect */
-			cbl_type = "USB";
-			cbl_state = CABLE_DETACHED;
-		} else {
-			pr_warn("DPM:%s:DR_SWAP cann't be processed\n",
-					__func__);
-			goto drole_err;
+			dpm_notify_cable_state(dpm, "USB-Host",
+							CABLE_DETACHED);
 		}
-		break;
-	case DATA_ROLE_UFP:
 		/* Send UFP connect */
 		cbl_type = "USB";
 		cbl_state = CABLE_ATTACHED;
 		break;
 
 	case DATA_ROLE_DFP:
+		if (cur_drole == DATA_ROLE_UFP) {
+			/* Role swap from UFP to DFP, Send UFP disconnect */
+			dpm_notify_cable_state(dpm, "USB",
+						CABLE_DETACHED);
+		}
 		/* Send DFP connect */
 		cbl_type = "USB-Host";
 		cbl_state = CABLE_ATTACHED;
@@ -476,8 +472,11 @@ static void dpm_update_data_role(struct devpolicy_mgr *dpm,
 		} else if (cur_drole == DATA_ROLE_UFP) {
 			cbl_type = "USB";
 			cbl_state = CABLE_DETACHED;
-		} else
+		} else {
+			pr_err("DMP:%s: Unexpected cur_drole=%d\n", __func__,
+						cur_drole);
 			goto drole_err;
+		}
 		break;
 	default:
 		pr_debug("DPM:%s: unknown data role!!\n", __func__);
-- 
1.9.1

